### ISSUE TITLE
Detekt 1.22.0-RC2 UnusedImports with type resolution false positive for androidx.core

### ISSUE THREAD

**Issue by eygraber** [detekt/detekt#5502]
This was not happening in 1.22.0-RC1

When running `detektDebug` in my Android project, I'm getting a bunch of `UnusedImports` false positives only for imports from `androidx.core.*` 

For example:

```
androidx.core.content.getSystemService
androidx.core.content.res.use
androidx.core.view.postDelayed
androidx.core.content.edit
```

**Comment by TWiStErRob** [comment]
Attempted repro (this passes with and without type resolution):
```kotlin
    @Test
    fun `does not report unused import for import used from androidX - #5502`() {
        val mainFile = """
            import android.content.Context
            
            fun f(context: Context) {
                val service: String = context.getSystemService()
            }
        """.trimIndent()

        val getSystemService = """
            package androidx.core.content
            import android.content.Context
            // https://developer.android.com/reference/kotlin/androidx/core/content/package-summary#(android.content.Context).getSystemService()
            inline fun <T : Any> Context.getSystemService(): T? = TODO()
        """.trimIndent()

        val androidContext = """
            package android.content
            class Context
        """.trimIndent()

        assertThat(subject.lintWithContext(env, mainFile, androidContext, getSystemService)).isEmpty()
    }
```

Can you please help out with the body of `f`? The usage style might be important.

**Comment by eygraber** [comment]
```
import android.content.Context
import android.view.LayoutInflater
import androidx.core.content.getSystemService

@Suppress("UnsafeCallOnNullableType")
inline val Context.layoutInflater: LayoutInflater
  get() = getSystemService()!!
```

```
import android.content.SharedPreferences
import androidx.core.content.edit
import kotlin.reflect.KProperty

typealias FieldNameFormatter = (KProperty<*>) -> String

interface SharedPreferencesDelegate<R, T> {
  operator fun getValue(thisRef: R, property: KProperty<*>): T
  operator fun setValue(thisRef: R, property: KProperty<*>, value: T)
}

inline fun <reified R, reified T> SharedPreferences.property(
  default: T,
  crossinline getter: SharedPreferences.(String, T) -> T,
  crossinline setter: SharedPrefsEditor.(String, T) -> Any,
  crossinline formatter: FieldNameFormatter = { it.name },
  crossinline postSetter: (T) -> Unit = {}
): SharedPreferencesDelegate<R, T> = object : SharedPreferencesDelegate<R, T> {
  override fun getValue(
    thisRef: R,
    property: KProperty<*>
  ): T = this@property.get(property, default, getter, formatter)

  override fun setValue(thisRef: R, property: KProperty<*>, value: T) {
    this@property.edit { put(property, value, setter, formatter) }
    postSetter(value)
  }
}
```

**Comment by TWiStErRob** [comment]
<details><summary>1st example still passes</summary>

```kotlin
    @Test
    fun `does not report unused import for import used from androidX 1 - #5502`() {
        val mainFile = """
            import android.content.Context
            import android.view.LayoutInflater
            import androidx.core.content.getSystemService
            
            inline val Context.layoutInflater: LayoutInflater
                get() = getSystemService()!!
        """.trimIndent()

        val getSystemService = """
            package androidx.core.content
            import android.content.Context
            // https://developer.android.com/reference/kotlin/androidx/core/content/package-summary#(android.content.Context).getSystemService()
            inline fun <T : Any> Context.getSystemService(): T? = TODO()
        """.trimIndent()

        val androidContext = """
            package android.content
            class Context
        """.trimIndent()

        val androidLayoutInflater = """
            package android.view
            class LayoutInflater
        """.trimIndent()

        assertThat(subject.lintWithContext(env, mainFile, androidContext, androidLayoutInflater, getSystemService))
            .isEmpty()
    }
```

</details>

<details><summary>and 2nd too</summary>

```kotlin
    @Test
    fun `does not report unused import for import used from androidX 2 - #5502`() {
        val mainFile = """
            import android.content.SharedPreferences
            import androidx.core.content.edit
            import kotlin.reflect.KProperty
            
            typealias FieldNameFormatter = (KProperty<*>) -> String
            
            interface SharedPreferencesDelegate<R, T> {
                operator fun getValue(thisRef: R, property: KProperty<*>): T
                operator fun setValue(thisRef: R, property: KProperty<*>, value: T)
            }
            
            inline fun <T> SharedPreferences.get(
                property: KProperty<*>,
                default: T,
                getter: SharedPreferences.(String, T) -> T,
                formatter: FieldNameFormatter
            ): T = TODO()
            
            inline fun <T> SharedPreferences.Editor.put(
                property: KProperty<*>,
                value: T,
                setter: SharedPreferences.Editor.(String, T) -> Any,
                formatter: (KProperty<*>) -> String
            ): Unit = TODO()
            
            inline fun <reified R, reified T> SharedPreferences.property(
                default: T,
                crossinline getter: SharedPreferences.(String, T) -> T,
                crossinline setter: SharedPreferences.Editor.(String, T) -> Any,
                crossinline formatter: FieldNameFormatter = { it.name },
                crossinline postSetter: (T) -> Unit = {}
            ): SharedPreferencesDelegate<R, T> = object : SharedPreferencesDelegate<R, T> {
                override fun getValue(thisRef: R, property: KProperty<*>): T =
                    this@property.get(property, default, getter, formatter)
            
                override fun setValue(thisRef: R, property: KProperty<*>, value: T) {
                    this@property.edit { put(property, value, setter, formatter) }
                    postSetter(value)
                }
            }
        """.trimIndent()

        val edit = """
            package androidx.core.content
            import android.content.SharedPreferences
            // https://developer.android.com/reference/kotlin/androidx/core/content/package-summary#(android.content.SharedPreferences).edit(kotlin.Boolean,kotlin.Function1)
            inline fun SharedPreferences.edit(
                commit: Boolean = false,
                action: SharedPreferences.Editor.() -> Unit
            ): Unit = TODO()
        """.trimIndent()

        val androidSharedPreferences = """
            package android.content
            interface SharedPreferences {
                interface Editor
            }
        """.trimIndent()

        assertThat(subject.lintWithContext(env, mainFile, androidSharedPreferences, edit))
            .isEmpty()
    }
```
</details>

To attack a different angle I looked at [RC2 diff](https://github.com/detekt/detekt/compare/v1.22.0-RC1...v1.22.0-RC2), but nothing stood out.

I think I'll defer to more experienced people in Detekt on this at this point.

**Comment by TWiStErRob** [comment]
Idea: turn on [`Detekt` or `DetektExtension` `debug`](https://detekt.dev/docs/gettingstarted/gradle/#kotlin-dsl-3) to see if there's relevant output.

**Comment by eygraber** [comment]
I'll give that a shot. I vaguely remember a classpath related change for RC2, so it might be that.

**Comment by eygraber** [comment]
When running with `debug` these errors are logged for the first example file:

```
error: unresolved reference: core
import androidx.core.content.getSystemService
                ^
error: type mismatch: inferred type is Any! but LayoutInflater? was expected
  get() = getSystemService()!!
          ^
error: no value passed for parameter 'p0'
  get() = getSystemService()!!

```

**Comment by TWiStErRob** [comment]
Repro project linked above. Steps:

 * Clone https://github.com/TWiStErRob/repros
 * open `detekt/UnusedImports-androidx.core-rc`
 * `gradlew cleanDetektDebug detektDebug`

Expected:
```
src\main\java\com\example\layoutInflater.kt:8:11: Calling !! on a nullable type will throw a NullPointerException at runtime in case the value is null. It should be avoided. [UnsafeCallOnNullableType]
```

Actual:
```
src\main\java\com\example\layoutInflater.kt:5:17: error: unresolved reference: core
import androidx.core.content.getSystemService
                ^
src\main\java\com\example\layoutInflater.kt:8:11: error: type mismatch: inferred type is Any! but LayoutInflater? was expected
  get() = getSystemService()!!
          ^
src\main\java\com\example\layoutInflater.kt:8:28: error: no value passed for parameter 'p0'
  get() = getSystemService()!!
                           ^

src\main\java\com\example\layoutInflater.kt:5:1: The import 'androidx.core.content.getSystemService' is unused. [UnusedImports]
```

@detekt this looks like a regression, because with the repro it's easy to confirm that it was indeed working in RC1. And it seems to highlight a bigger issue than just `UnusedImports`.
